parsing/location.ml

```diff
@@ -765,6 +770,53 @@ let batch_mode_printer : report_printer =
   { pp; pp_report_kind; pp_main_loc; pp_main_txt;
     pp_submsgs; pp_submsg; pp_submsg_loc; pp_submsg_txt }
 
+
+let json_mode_printer: json_report_printer = fun json report ->
+  let kind = match report.kind with
+    | Report_error -> `Assoc["kind", `String "error" ]
+    | Report_warning w ->
+        `Assoc ["kind", `String "warning"; "id", `String w]
+    | Report_warning_as_error w ->
+        `Assoc ["kind",`String "error_from_warning"; "id", `String w]
+    | Report_alert w ->
+        `Assoc ["kind",`String "alert"; "id",`String w]
+    | Report_alert_as_error w ->
+        `Assoc ["kind",`String "error_from_alert"; "id", `String w]
+  in
+  let content txt = `String (Format.asprintf "@[%t@]" txt) in
+  let loc_to_json loc =
+    let file =
+      if loc.loc_start.pos_fname = "" then !input_name
+      else loc.loc_start.pos_fname
+    in
+    let startchar = loc.loc_start.pos_cnum - loc.loc_start.pos_bol in
+    let endchar = loc.loc_end.pos_cnum - loc.loc_end.pos_bol in
+    let startline = loc.loc_start.pos_lnum in
+    let endline = loc.loc_end.pos_lnum in
+    let start_end s e = `List [`Int s; `Int e] in
+    `Assoc[
+      "file", `String file;
+      "lines", start_end startline endline;
+      "characters", start_end startchar endchar;
+    ]
+  in
+  let msg_to_json {loc;txt} =
+    `Assoc[
+      "location", loc_to_json loc;
+      "content", content txt;
+    ]
+  in
+  let submsgs = List.map msg_to_json report.sub in
+  let main = msg_to_json report.main in
+  let err_frag = `Assoc[
+      "info", kind;
+      "msg", main;
+      "submsgs",`List(submsgs);
```

Comment:
Another naming choice would be to use `"main"` and `"sub"` here to follow the OCaml type, and then `"msg"` as the key that is currently `"content"` in `msg_to_json`. I don't have strong opinions here.